ntoskrnl/include/internal/i386/asmmacro.S

```diff
@@ -249,19 +249,24 @@ MACRO(TRAP_ENTRY, Trap, Flags)
     .ENDP
 ENDM
 
+#define KI_NMI  HEX(0001)
+
 MACRO(TASK_ENTRY, Trap, Flags)
-    // EXTERN @&Trap&Handler@0 :PROC
     EXTERN _&Trap&Handler :PROC
     PUBLIC _&Trap
     .PROC _&Trap
         /* Generate proper debugging symbols */
         FPO 0, 0, 0, 0, 0, FRAME_TSS
 
-        // /* Common code to create the trap frame */
-        // KiEnterTrap Flags
-
         /* Call the C handler */
-        KiCallHandler _&Trap&Handler // @&Trap&Handler@0
+        call _&Trap&Handler
+
+        if (Flags AND KI_NMI)
+            /* Return from NMI: return with iret and handle NMI recursion */
+            iretd
+            jmp _&Trap
+        endif
```

Comment:
See also https://github.com/minoca/os/blob/5bb001e90f1f5fe237a9664f23eaee3cb2e57d22/kernel/x86/trap.S#L330
and https://github.com/minoca/os/blob/5bb001e90f1f5fe237a9664f23eaee3cb2e57d22/kernel/x86/prochw.c#L1005
from MinocaOS project.

https://git.reactos.org/?p=reactos.git;a=blob;f=reactos/ntoskrnl/ke/i386/trap.s;h=4fa3f5b90129be69f548b783a3556be48837857d;hb=0940d6e34ab59594d70be744f9a00dd801f07dae#l673
may be worth it (for NMI).

List of things I need to recheck: https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=commit&s=Trap+Handlers